src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelModuleResolutionFunction.java

```diff
@@ -61,36 +66,89 @@ public SkyValue compute(SkyKey skyKey, Environment env)
     if (root == null) {
       return null;
     }
-    ImmutableMap<ModuleKey, InterimModule> initialDepGraph = Discovery.run(env, root);
+
+    var state = env.getState(ModuleResolutionComputeState::new);
+    if (state.selectionResult == null) {
+      state.selectionResult = discoverAndSelect(env, root);
+      if (state.selectionResult == null) {
+        return null;
+      }
+    }
+
+    ImmutableSet<RepoSpecKey> repoSpecKeys =
+        state.selectionResult.getResolvedDepGraph().values().stream()
+            // Modules with a null registry have a non-registry override. We don't need to
+            // fetch or store the repo spec in this case.
+            .filter(module -> module.getRegistry() != null)
+            .map(RepoSpecKey::of)
+            .collect(toImmutableSet());
+    SkyframeLookupResult repoSpecResults = env.getValuesAndExceptions(repoSpecKeys);
+    ImmutableMap.Builder<ModuleKey, RepoSpec> remoteRepoSpecs = ImmutableMap.builder();
+    for (RepoSpecKey repoSpecKey : repoSpecKeys) {
+      RepoSpec repoSpec = (RepoSpec) repoSpecResults.get(repoSpecKey);
+      if (repoSpec == null) {
+        return null;
+      }
+      remoteRepoSpecs.put(repoSpecKey.getModuleKey(), repoSpec);
+    }
+
+    ImmutableMap<ModuleKey, Module> finalDepGraph;
+    try (SilentCloseable c =
+        Profiler.instance().profile(ProfilerTask.BZLMOD, "compute final dep graph")) {
+      finalDepGraph =
+          computeFinalDepGraph(
+              state.selectionResult.getResolvedDepGraph(),
+              root.getOverrides(),
+              remoteRepoSpecs.buildOrThrow());
+    }
+
+    Profiler.instance().profile(ProfilerTask.BZLMOD, "module resolution completed").close();
```

Comment:
nit -- is this necessary? isn't the end of the 'compute final dep graph' exactly this?